src/lib/IDBDAO.js

```diff
@@ -0,0 +1,350 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'IDBInternalException',
+  extends: 'foam.dao.InternalException',
+
+  // TODO: Which errors are internal (system problems) vs. external
+  // (i.e. invalid data for clone, but you can try again with different data)
+  properties: [
+    'id',
+    'error',
+    {
+      name: 'message',
+      expression: function(id, error) {
+        return "IndexedDB Error for " + id +
+          ( error ? ": " + error.toString() : "" );
+      }
+    }
+  ]
+});
+
+
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'IDBDAO',
+  extends: 'foam.dao.AbstractDAO',
+
+  requires: [
+    'foam.dao.FlowControl',
+    'foam.dao.ArraySink',
+    'foam.dao.IDBInternalException',
+    'foam.mlang.predicate.True',
+    'foam.mlang.predicate.Eq',
+  ],
+
+  imports: [
+    'async',
+  ],
+
+  constants: {
+    /** Global cache of the current transaction reference. Only element 0 is used. */
+    __TXN__: [],
+  },
+
+  properties: [
+    {
+      name:  'of',
+      required: true
+    },
+    {
+      name:  'name',
+      label: 'Store Name',
+      expression: function() {
+        return this.of.id;
+      }
+    },
+    {
+      name: 'indicies',
+      factory: function() { return []; }
+    },
+    {
+      /** @internal */
+      name: 'propKeys_'
+    },
+    {
+      /** The promise that holds the open DB. Call this.withDB.then(function(db) { ... }); */
+      name: 'withDB',
+      factory: function() {
+        var self = this;
+        return new Promise(function(resolve, reject) {
+          var indexedDB = window.indexedDB ||
+            window.webkitIndexedDB         ||
+            window.mozIndexedDB;
+
+          var request = indexedDB.open("FOAM:" + self.name, 1);
+
+          request.onupgradeneeded = function(e) {
+            var store = e.target.result.createObjectStore(self.name);
+            for ( var i = 0; i < self.indicies.length; i++ ) {
+              store.createIndex(self.indicies[i][0], self.indicies[i][0], { unique: self.indicies[i][1] });
+            }
+          }
+
+          request.onsuccess = function(e) {
+            resolve(e.target.result);
+          }
+
+          request.onerror = function (e) {
+            reject(self.IDBInternalException.create({ id: 'open', error: e }));
+          };
+        });
+      }
+    },
+  ],
+
+  methods: [
+
+    function init() {
+
+      var propKeys = this.propKeys_ = {};
```

Comment:
Probably safe to delete this.  It looks like propKeys_ was only used in the FOAM1 IDBDAO for simple serialization.  But we should just make foam.json smart enough to use shortnames if available.